stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+        skipWhiteSpace(parserState)
+
+        let c = parserState.currentCodePoint
+        match (c) {
+          -1 => {
+            let detail = "expected a key-value pair or the end of the object"
+            err = Some(buildUnexpectedTokenError(parserState, detail))
+            break
+          },
+          0x2C => { // ','
+            trailingComma = true
+            if (first) {
+              let detail = "expected a key-value pair or the end of the object, found ','"
+              err = Some(buildUnexpectedTokenError(parserState, detail))
+              break
+            } else {
+              next(parserState)
+              void
+            }
+          },
+          0x7D => { // '}'
+            if (trailingComma) {
+              let detail = "unexpected trailing comma in object"
+              err = Some(buildUnexpectedTokenError(parserState, detail))
+              break
+            }
+            next(parserState)
+            done = true
+            break
+          },
+          _ => {
+            trailingComma = false
+            // A new entry in current object.
+            // Just call parseString directly. In case the current character id not '"', it will return an error we can pass along.
+            match (parseString(parserState)) {
+              Ok(key) => {
+                skipWhiteSpace(parserState)
+
+                match (expectCodePointAndAdvance(0x3A, parserState)) {
+                  // ':'
+                  None => {
+                    // note that parseValue skips initial and final whitespace
+                    match (parseValue(parserState)) {
+                      Ok(value) => {
+                        entries = [(key, value), ...entries]
+                        first = false
+                        void
+                      },
+                      Err(e) => {
+                        err = Some(e)
+                        break
+                      },
+                    }
+                  },
+                  Some(e) => {
+                    err = Some(e)
+                    break
+                  },
+                }
+              },
+              Err(e) => {
+                err = Some(e)
+                break
+              },
+            }
+          },
+        }
+      }
+      // end of entry loop
+
+      match (err) {
+        Some(e) => Err(e),
+        None => {
+          if (done) {
+            Ok(JsonObject(List.reverse(entries)))
+          } else {
+            // This branch is not expected to actually execute,
+            // but in case it does, may just as well do the right thing.
+            Err(
+              buildUnexpectedTokenError(parserState, "unexpected end of object")
+            )
+          }
+        },
+      }
+    },
+  }
+}
+
+/**
+ * Parses JSON input from a string into a `JSON` object.
+ *
+ * @param str: The JSON text string
+ * @returns A `Result` object with either the parsed `JSON` object or an error.
```

Comment:
```suggestion
 * @returns The parsed `JSON` data structure if successful or the problem otherwise
```

Reply:
This doesnt really match the pattern we use throughout the stdlib for documenting `Results`, I feel like the original might need to be changed to ```
`Ok(value)` containing the parsed Json datastructure if successful otherwise `Err(err)` containing the corresponding parse error.
or something like that